node/core/dispute-coordinator/src/db/v1.rs

```diff
@@ -0,0 +1,436 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! V1 database for the dispute coordinator.
+
+use polkadot_primitives::v1::{
+	CandidateReceipt, ValidDisputeStatementKind, InvalidDisputeStatementKind, ValidatorIndex,
+	ValidatorSignature, SessionIndex, CandidateHash,
+};
+
+use kvdb::{KeyValueDB, DBTransaction};
+use parity_scale_codec::{Encode, Decode};
+
+const ACTIVE_DISPUTES_KEY: &[u8; 15] = b"active-disputes";
+const EARLIEST_SESSION_KEY: &[u8; 16] = b"earliest-session";
+const CANDIDATE_VOTES_SUBKEY: &[u8; 15] = b"candidate-votes";
+
+// It would be nice to draw this from the chain state, but we have no tools for it right now.
+// On Polkadot this is 2 days, and on Kusama it's 12 hours.
+const DISPUTE_WINDOW: SessionIndex = 12;
+
+fn candidate_votes_key(session: SessionIndex, candidate_hash: &CandidateHash) -> [u8; 15 + 4 + 32] {
+	let mut buf = [0u8; 15 + 4 + 32];
+	buf[..15].copy_from_slice(CANDIDATE_VOTES_SUBKEY);
+
+	// big-endian encoding is used to ensure lexicographic ordering.
+	buf[15..][..4].copy_from_slice(&session.to_be_bytes());
+	candidate_hash.using_encoded(|s| buf[(15 + 4)..].copy_from_slice(s));
+
+	buf
+}
+
+// Computes the upper lexicographic bound on DB keys for candidate votes with a given
+// upper-exclusive bound on sessions.
+fn candidate_votes_range_upper_bound(upper_exclusive: SessionIndex) -> [u8; 15 + 4] {
+	let mut buf = [0; 15 + 4];
+	buf[..15].copy_from_slice(CANDIDATE_VOTES_SUBKEY);
+	// big-endian encoding is used to ensure lexicographic ordering.
+	buf[15..][..4].copy_from_slice(&upper_exclusive.to_be_bytes());
+
+	buf
+}
+
+fn decode_candidate_votes_key(key: &[u8]) -> Option<(SessionIndex, CandidateHash)> {
+	if key.len() != 15 + 4 + 32 {
+		return None;
+	}
+
+	let mut session_buf = [0; 4];
+	session_buf.copy_from_slice(&key[15..][..4]);
+	let session = SessionIndex::from_be_bytes(session_buf);
+
+	CandidateHash::decode(&mut &key[(15 + 4)..]).ok().map(|hash| (session, hash))
+}
+
+/// Column configuration information for the DB.
+#[derive(Debug, Clone)]
+pub struct ColumnConfiguration {
+	/// The column in the key-value DB where data is stored.
+	pub col_data: u32,
+}
+
+/// Tracked votes on candidates, for the purposes of dispute resolution.
+#[derive(Debug, Clone, Encode, Decode)]
+pub struct CandidateVotes {
```

Comment:
Duplicate of primitives.

Reply:
I believe that's intentional. We want to separate DB layout from the in-memory type definition, same as in the approval-voting DB.

Reply:
Yep, that's the intent. I will be adding conversions in follow-on commits.